File: neet_code_all_cpp/math_and_geometry/matrix_diagonal_sum.cpp
#include <iostream>
#include <vector>

class Solution {
    public:
        // AC
        int diagonalSum(std::vector<std::vector<int>>& mat) {
            int n = mat.size();
            int result = 0;
            for(int i = 0; i < n; ++i) {
                result += mat[i][i];
                if(i != n - 1 - i) {
                    result += mat[i][n - 1 - i];
                }
            }

            return result;
        }
};

// Kotlinの模範解答
class SolutionAnsKotlin {
    public:
        int diagonalSum(std::vector<std::vector<int>>& mat) {
            int n = mat.size() - 1;
            int result = 0;

            for(int i = 0; i <= n; ++i) {
                result += mat[i][i] + mat[n - i][i];
            }

            if (n % 2 == 0){
                result -= mat[n/2][n/2];
            }

            return result;
        }
};

int main(void) {
    std::vector<std::vector<int>> case_1 = {{1,2,3},
        {4,5,6},
        {7,8,9}};
    // => 25
    std::vector<std::vector<int>> case_2 = {{1,1,1,1},
        {1,1,1,1},
        {1,1,1,1},
        {1,1,1,1}};
    // => 8
    std::vector<std::vector<int>> case_3 = {{5}};
    // => 5

    Solution s_1;

    std::cout << s_1.diagonalSum(case_1) << std::endl;
    std::cout << s_1.diagonalSum(case_2) << std::endl;
    std::cout << s_1.diagonalSum(case_3) << std::endl;

    SolutionAnsKotlin s_ans_kt;

    std::cout << s_ans_kt.diagonalSum(case_1) << std::endl;
    std::cout << s_ans_kt.diagonalSum(case_2) << std::endl;
    std::cout << s_ans_kt.diagonalSum(case_3) << std::endl;
}

File: neet_code_all/src/math_and_geometry/matrix_diagonal_sum.rs
struct Solution;
impl Solution {
    // AC(JSの模範解答と全く同じだった)
    // (i,i) と (i, n - 1 - i)が等しい場合は除くことでダブルカウントを防ぐ
    pub fn diagonal_sum(mat: Vec<Vec<i32>>) -> i32 {
        let mut result = 0;
        let n = mat.len();
        for i in 0..n {
            result += mat[i][i];
            if i != n - 1 - i {
                result += mat[i][n - 1 - i];
            }
        }

        result
    }
}

// Kotlinの模範解答より
struct SolutionAnsKotlin;
impl SolutionAnsKotlin {
    pub fn diagonal_sum(mat: Vec<Vec<i32>>) -> i32 {
        let n = mat.len() - 1;
        let mut result = 0;
        for i in 0..=n {
            result += mat[i][i] + mat[n - i][i];
        }

        // nが偶数のときは(n/2, n/2)をダブルカウントするのであとから引き算しておく
        if n % 2 == 0 {
            result -= mat[n / 2][n / 2];
        }

        result
    }
}

fn main() {
    let case_1 = vec![vec![1, 2, 3], vec![4, 5, 6], vec![7, 8, 9]];
    // => 25
    let case_2 = vec![
        vec![1, 1, 1, 1],
        vec![1, 1, 1, 1],
        vec![1, 1, 1, 1],
        vec![1, 1, 1, 1],
    ];
    // => 8
    let case_3 = vec![vec![5]];
    // =>5

    println!("case_1: {}", Solution::diagonal_sum(case_1.clone()));
    println!("case_2: {}", Solution::diagonal_sum(case_2.clone()));
    println!("case_3: {}", Solution::diagonal_sum(case_3.clone()));

    println!(
        "case_1: {}",
        SolutionAnsKotlin::diagonal_sum(case_1.clone())
    );
    println!(
        "case_2: {}",
        SolutionAnsKotlin::diagonal_sum(case_2.clone())
    );
    println!(
        "case_3: {}",
        SolutionAnsKotlin::diagonal_sum(case_3.clone())
    );
}
